### ISSUE TITLE
cmus does not properly detect width of Japanese characters

### ISSUE THREAD

**Issue by wooosh** [cmus/cmus#1196]
On cmus version 2.9.1-25-g54d7213 (current commit at the time of this issue's creation), cmus is not able to determine the width of Japanese characters accurately, resulting in broken right alignment:

![image](https://user-images.githubusercontent.com/12696312/169419762-796825bd-8069-41ac-baf5-3b797de9b5e5.png)
![image](https://user-images.githubusercontent.com/12696312/169419790-2dfefbf7-1475-48ab-bf6d-89e55a6fa246.png)

Replacing the `u_char_width` implementation with a call to POSIX `wcwidth` using glibc 2.32.2 appears to fix the problem:

```
int u_char_width(uchar u) {
        return wcwidth(u);
}
```

![image](https://user-images.githubusercontent.com/12696312/169420214-4c95372f-abf7-4719-a350-ea7c664bd144.png)
![image](https://user-images.githubusercontent.com/12696312/169420239-92a8bab6-8e56-46d2-997f-454e95842ee5.png)

Unfortunately, I think this does not actually solve the problem completely, as different libcs provide different implementations, and glibc's `wcwidth` was not up to date for a number of years IIRC. 

I'd be more than happy to submit a PR if necessary, but I have two concerns:

* There are other issues open relating to character spacing, and it's unclear to me if this code will be replaced in the near future regardless
* The wcwidth implementations in musl and glibc are both table based, so it's unclear to me if it would be preferred to fix this by switching `u_char_width` to a table based approach
* There is also the possibility of vendoring the musl implementation, which is relatively concise, and seems like it would require less maintenance

**Comment by gavtroy** [comment]
It seems to me it would be fine, although I do wonder if there's a reason it wasn't done that way to start with. I had also experimented with wcwidth but chickened out because
- I was confused about it typically being locale aware, but judging from your experiment that's actually exactly what we want.
- Windows. Might need to fall back to the existing implementation there since wchar_t is only 2 bytes.
- uchar is a unicode index, while wchar_t is _technically_ implementation defined. A vendored implementation would address this, but I'd be inclined to just ignore it as a purely theoretical issue.

> There are other issues open relating to character spacing, and it's unclear to me if this code will be replaced in the near future regardless

Nah, #1151 adds a small hack to fix some emoji characters but "ultimately this is not a solution but just an easy short term improvement over what we have". It also changes how combining characters are treated but that's something we should change regardless of whether we use wcwidth.

What locale do you use? And (asked in ignorance) is there a reason you use it over a UTF-8 locale?

**Comment by wooosh** [comment]
Here is the output of `locale`:

```
LANG=en_US.UTF-8
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_COLLATE=C
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_PAPER="en_US.UTF-8"
LC_NAME="en_US.UTF-8"
LC_ADDRESS="en_US.UTF-8"
LC_TELEPHONE="en_US.UTF-8"
LC_MEASUREMENT="en_US.UTF-8"
LC_IDENTIFICATION="en_US.UTF-8"
LC_ALL=
```

I believe I am using the UTF-8 locale. It's worth noting that most of the songs in my library with Japanese titles have the correct width, just not all of them.

Here are some of the problematic song titles in plaintext if you'd like to try to reproduce it:

```
ゼロコンマ、色とりどりの世界
まで。情熱ミックス
終わりのはじまり
```

**Comment by wooosh** [comment]
Looking through the source code and documentation of (u)rxvt and (u)xterm, both rely on the system implementation of `wcwidth` if available.

urxvt: http://pod.tst.eu/http://cvs.schmorp.de/rxvt-unicode/doc/rxvt.7.pod#Character_widths_are_not_correct

uxterm: https://github.com/joejulian/xterm/blob/defc6dd5684a12dc8e56cb6973ef973e7a32caa3/util.c#L70

This leads me to believe that `wcwidth` is likely more correct than vendoring a solution on systems with a correctly sized `wchar_t`, because cmus should agree with the terminal, even if the terminal is using a weird locale or outdated implementation of `wcwidth`.

**Comment by gavtroy** [comment]
So not locale, excuse my confusion there. In fact wcwidth doesn't seem to affect how broken we are on unusual locales.

I can't reproduce it just by copy paste but I guess in your files the likes of ゼ encodes the dakuten separately as combining mark [U+3099](https://util.unicode.org/UnicodeJsps/character.jsp?a=3099). In fact it looks like we have code specifically for such characters but it does the wrong thing, e.g.
```
    if (u == 0x3099U)
        goto narrow;
```
I'd expect them to be treated as zero-wide rather than narrow. We should fix the existing code anyway if it's being kept as a fallback, but wcwidth will fix many other edge cases so it's still worth proposing. You might get more attention with a pull request (although it doesn't look like anyone is in a rush :slightly_smiling_face:).

**Comment by gavtroy** [comment]
You might also have noticed that this unusually-encoded ゼ is displayed as セ, and similar for other characters. This turns out to be a separate issue to us calculating the wrong width. I reported it upstream to ncurses and in fact Thomas E. Dickey has already implemented a fix in patch 20220612.
